consensus/consensus-types/src/proposal_msg.rs

```diff
@@ -26,8 +26,16 @@ impl ProposalMsg {
         }
     }
 
-    pub fn epoch(&self) -> u64 {
-        self.proposal.epoch()
+    /// Returns the epoch associated with the proposal after verifying that the
+    /// payload data is also associated with the same epoch
+    pub fn verified_epoch(&self) -> Result<u64> {
```

Comment:
it seems better to verify the payload epoch in verify_well_formed function and keep here to just return u64?

Reply:
this is how we did for BatchMsg seems like, copied that. moved it to `verify_well_formed`.